File: simplified-books-core/src/main/java/org/nypl/simplified/books/core/BookDatabaseEntrySnapshot.java
package org.nypl.simplified.books.core;

import com.io7m.jfunctional.OptionType;
import com.io7m.jnull.NullCheck;
import org.nypl.drm.core.AdobeAdeptLoan;
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntry;

import java.io.File;

/**
 * A snapshot of the most recent state of a book in the database.
 */

public final class BookDatabaseEntrySnapshot
{
  private final BookID                     id;
  private final OptionType<AdobeAdeptLoan> adobe_rights;
  private final OptionType<File>           book;
  private final OptionType<File>           cover;
  private final OPDSAcquisitionFeedEntry   entry;

  /**
   * Construct a book snapshot.
   *
   * @param in_id           The book ID
   * @param in_cover        The cover file, if any
   * @param in_book         The actual book (typically an EPUB), if any
   * @param in_entry        The acquisition feed entry
   * @param in_adobe_rights The Adobe DRM rights associated with the book, if
   *                        any
   */

  public BookDatabaseEntrySnapshot(
    final BookID in_id,
    final OptionType<File> in_cover,
    final OptionType<File> in_book,
    final OPDSAcquisitionFeedEntry in_entry,
    final OptionType<AdobeAdeptLoan> in_adobe_rights)
  {
    this.id = NullCheck.notNull(in_id);
    this.cover = NullCheck.notNull(in_cover);
    this.book = NullCheck.notNull(in_book);
    this.entry = NullCheck.notNull(in_entry);
    this.adobe_rights = NullCheck.notNull(in_adobe_rights);
  }

  /**
   * @return The book file (typically an EPUB), if any
   */

  public OptionType<File> getBook()
  {
    return this.book;
  }

  /**
   * @return The book ID
   */

  public BookID getBookID()
  {
    return this.id;
  }

  /**
   * @return The cover image, if any
   */

  public OptionType<File> getCover()
  {
    return this.cover;
  }

  /**
   * @return The Adobe DRM rights, if any
   */

  public OptionType<AdobeAdeptLoan> getAdobeRights()
  {
    return this.adobe_rights;
  }

  /**
   * @return The acquisition feed entry
   */

  public OPDSAcquisitionFeedEntry getEntry()
  {
    return this.entry;
  }

  @Override public String toString()
  {
    final StringBuilder sb = new StringBuilder("BookDatabaseEntrySnapshot{");
    sb.append("adobe_rights=").append(this.adobe_rights);
    sb.append(", id=").append(this.id);
    sb.append(", book=").append(this.book);
    sb.append(", cover=").append(this.cover);
    sb.append(", entry=").append(this.entry.getAvailability());
    sb.append('}');
    return sb.toString();
  }
}

File: simplified-books-core/src/main/java/org/nypl/simplified/books/core/BookDatabaseEntrySnapshot.kt
package org.nypl.simplified.books.core

import com.io7m.jfunctional.Option
import com.io7m.jfunctional.OptionType
import org.nypl.simplified.books.core.BookDatabaseEntryFormatSnapshot.BookDatabaseEntryFormatSnapshotAudioBook
import org.nypl.simplified.books.core.BookDatabaseEntryFormatSnapshot.BookDatabaseEntryFormatSnapshotEPUB
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntry
import java.io.File

/**
 * A snapshot of the most recent state of a book in the database.
 */

data class BookDatabaseEntrySnapshot(

  /**
   * @return The book ID
   */

  val bookID: BookID,

  /**
   * @return The cover image, if any
   */

  val cover: OptionType<File>,

  /**
   * @return The acquisition feed entry
   */

  val entry: OPDSAcquisitionFeedEntry,

  /**
   * @return A snapshot of all formats within the database entry
   */

  val formats: List<BookDatabaseEntryFormatSnapshot>) {

  /**
   * If any format is downloaded, then the book as a whole is currently considered to be downloaded
   */

  val isDownloaded: Boolean
    get() = this.formats.any { format -> format.isDownloaded }

  /**
   * @return The format of the given type, if one is present
   */

  fun <T : BookDatabaseEntryFormatSnapshot> findFormat(clazz: Class<T>): OptionType<T> {
    return Option.of(this.formats.find { format -> clazz.isAssignableFrom(format.javaClass) } as T?)
  }

  /**
   * @return The "preferred" format for the given type, if any
   */

  fun findPreferredFormat(): OptionType<BookDatabaseEntryFormatSnapshot> {
    val formats = mutableListOf<BookDatabaseEntryFormatSnapshot>()
    formats.addAll(this.formats.filterIsInstance(
      BookDatabaseEntryFormatSnapshotEPUB::class.java))
    formats.addAll(this.formats.filterIsInstance(
      BookDatabaseEntryFormatSnapshotAudioBook::class.java))
    formats.addAll(this.formats)
    return Option.of(formats.firstOrNull())
  }
}
